### PR TITLE
Add breadcrumb trail to messages

### LINKED ISSUE #771: Breadcrumb metadata for broadcasted messages (blocks and txs)
In order to investigate the path each message had come from so that we can be aware of each node's stay time.  This could be added as a field to the message header, like message signature, and the field should encode a sequence of (node id, departed timestamp) pairs.  For the precision, the departed timestamp would be better to include milliseconds or even microseconds.

### PR DESCRIPTION
**Pull request by limebell**
This patch adds metadata for breadcrumb trail to messages.

Todo:
- ~Add unit tests~
- ~Add viewer~

Closes #771

### CODE REVIEW

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 1]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+                swarmC.Dispose();
+                swarmD.Dispose();
+            }
+        }
+
+        [Fact(Timeout = Timeout)]
+        public async Task BreadcrumbTrail()
+        {
+            var keys = new PrivateKey[5];
+            var fxs = new DefaultStoreFixture[5];
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
            var fxs = new DefaultStoreFixture[keys.Length];
```

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 2]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+            var blockChains = new BlockChain<DumbAction>[5];
+            var swarms = new Swarm<DumbAction>[5];
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
            var swarms = new Swarm<DumbAction>[fxs.Length];
```

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 3]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+                0x55, 0xd7, 0x38, 0xa3, 0x67, 0x48, 0x73, 0x08, 0xcd, 0x74,
+                0x3c, 0x4b,
+            });
+
+            var genesisBlock = BlockChain<DumbAction>.MakeGenesisBlock();
+
+            try
+            {
+                var tasks = new List<Task>();
+                for (int i = 0; i < 5; i++)
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
                for (int i = 0; i < fxs.Length; i++)
```

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 4]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+                swarmD.Dispose();
+            }
+        }
+
+        [Fact(Timeout = Timeout)]
+        public async Task BreadcrumbTrail()
+        {
+            var keys = new PrivateKey[5];
+            var fxs = new DefaultStoreFixture[5];
+            var blockChains = new BlockChain<DumbAction>[5];
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
            var blockChains = new BlockChain<DumbAction>[keys.Length];
```

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 5]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+                        blockChains[i],
+                        keys[i],
+                        0,
+                        host: IPAddress.Loopback.ToString());
+                    tasks.Add(StartAsync(swarms[i]));
+                }
+
+                await Task.WhenAll(tasks.ToArray());
+
+                for (int i = 0; i < 4; i++)
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
                for (int i = 0; i < swams.Length - 1; i++)
```

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 6]
```diff
@@ -2682,6 +2682,141 @@ public async Task FindSpecificPeerAsync()
...
+                expected += $"[{receivedTrail}:{swarms[4].Address}:" +
+                            $"{header.Trail.Trails[count].Timestamp}]";
+
+                Assert.Equal(
+                    expected,
+                    header.Trail.ToString());
+            }
+            finally
+            {
+                for (int i = 0; i < 5; i++)
```
Comment (dahlia) [inline Libplanet.Tests/Net/SwarmTest.cs]:
```suggestion
                for (int i = 0; i < swarms.Length; i++)
```

File: Libplanet/FixedSizedQueue.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace Libplanet
+{
+    /// <summary>
+    /// Implements a <see cref="ConcurrentQueue{T}"/> of a fixed size.
+    /// </summary>
+    /// <typeparam name="T">Specifies the type of elements in the queue.</typeparam>
+#pragma warning disable MEN002
+    // This class referenced the implementation of the following.
+    // https://github.com/couchcoding/Logbert/blob/bbed28fabf9493310f4b51324f8af88c14551c9d/src/Logbert/Helper/FixedSizedQueue.cs
```
Comment (dahlia) [inline Libplanet/FixedSizedQueue.cs]: Rather than suppressing the `MEN002` warning, how about shortening the link using [gio.io](https://git.io/), GitHub's official shortener?

File: Libplanet/FixedSizedQueue.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+#pragma warning restore MEN002
+    internal class FixedSizedQueue<T> : ConcurrentQueue<T>
+    {
+        /// <summary>
+        /// Simple object for thread synchronization.
```
Comment (dahlia) [inline Libplanet/FixedSizedQueue.cs]:
```suggestion
        /// Equivalent to <see cref="ConcurrentQueue{T}"/>, except this does not accept more than
        /// the specified maximum size.
```

Comment (limebell) [reply Libplanet/FixedSizedQueue.cs]: Maybe review of line 6?

Comment (dahlia) [reply Libplanet/FixedSizedQueue.cs]: Yeah, it was misplaced.

File: Libplanet.Tests/TestUtils.cs [hunk 9]
```diff
@@ -21,6 +21,82 @@ public class TestUtils
...
+                0x59, 0x79,
+            }),
+            new PrivateKey(new byte[]
+            {
+                0x3d, 0xae, 0x2d, 0x2f, 0x87, 0x1b, 0x11, 0xaa, 0x41, 0xbd,
+                0xec, 0x81, 0x4c, 0x4d, 0x27, 0xf3, 0xba, 0xd9, 0x1f, 0x61,
+                0xc3, 0x57, 0xab, 0x43, 0xa0, 0x0c, 0x63, 0x1b, 0x2b, 0x15,
+                0x13, 0xf2,
+            }),
+        };
```
Comment (dahlia) [inline Libplanet.Tests/TestUtils.cs]: :+1:

File: Libplanet/Net/Messages/BreadcrumbTrail.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            return Record
+                ? _trails.Aggregate(
+                    string.Empty,
+                    (s, trail) => s + $"[{trail}]/",
+                    s => s.TrimEnd('/'))
```
Comment (dahlia) [inline Libplanet/Net/Messages/BreadcrumbTrail.cs]: Can't this use `string.Join()` method instead?

File: Libplanet.Tests/Net/SwarmTest.cs [hunk 11]
```diff
@@ -2757,6 +2757,105 @@ public async Task FindSpecificPeerAsync()
...
+                swarmB.Dispose();
+                swarmC.Dispose();
+                swarmD.Dispose();
+            }
+        }
+
+        [Fact(Timeout = Timeout)]
+        public async Task BreadcrumbTrail()
+        {
+            int swarmCount = 5;
```
Comment (moreal) [inline Libplanet.Tests/Net/SwarmTest.cs]: It seems better to be const variable because `swarmCount` does not change.

```suggestion
            const int swarmCount = 5;
```

File: Libplanet/Net/Messages/GetRecentStates.cs [hunk 12]
```diff
@@ -12,19 +12,19 @@ internal class GetRecentStates : Message
...
         }
 
         public GetRecentStates(NetMQFrame[] frames)
-            : this(
-                new BlockLocator(frames.Skip(2).Select(f => new HashDigest<SHA256>(f.Buffer))),
-                new HashDigest<SHA256>(frames[1].Buffer),
-                frames[0].ConvertToInt64()
-            )
         {
+            BaseLocator =
```
Comment (longfin) [inline Libplanet/Net/Messages/GetRecentStates.cs]: Is this change necessary?

File: Libplanet/Net/Messages/Message.cs [hunk 13]
```diff
@@ -11,6 +11,19 @@ namespace Libplanet.Net.Messages
 {
     internal abstract class Message
     {
+        internal const int HeaderSize = 4;
+
+        protected const bool LeaveTrail = true;
```
Comment (longfin) [inline Libplanet/Net/Messages/Message.cs]: I guess it's okay that we omit this constant and use `true` or `false` directly. if we want to show its meaning, `base(trail: true)` is enough.

File: Libplanet/Net/Protocols/KademliaProtocol.cs [hunk 14]
```diff
@@ -816,12 +816,9 @@ private void ReceiveFindPeer(FindNeighbors findNeighbors)
         {
             IEnumerable<BoundPeer> found = _routing.Neighbors(findNeighbors.Target, _bucketSize);
 
-            Neighbors neighbors = new Neighbors(found)
-            {
-                Identity = findNeighbors.Identity,
-            };
+            Neighbors neighbors = new Neighbors(found);
```
Comment (longfin) [inline Libplanet/Net/Protocols/KademliaProtocol.cs]:
```suggestion
            var neighbors = new Neighbors(found);
```

File: Libplanet/Net/Protocols/RoutingTable.cs [hunk 15]
```diff
@@ -91,7 +91,9 @@ public IEnumerable<BoundPeer> PeersToBroadcast(Address? except)
             if (count < _minPeersToBroadcast)
             {
                 peers.AddRange(Peers
-                    .Where(peer => except is null || !peer.Address.Equals(except.Value))
+                    .Where(peer =>
+                        !peers.Contains(peer) &&
```
Comment (longfin) [inline Libplanet/Net/Protocols/RoutingTable.cs]: It doesn't seem related to adding breadcrumb. Is it intended? (if so, how about writing about it to changelogs?)

File: Libplanet/Net/Messages/GetRecentStates.cs (hunk 12 above)
Comment (limebell) [reply Libplanet/Net/Messages/GetRecentStates.cs]: Yes, because message generated with `NetMQFrame`s should keep previous breadcrumb trail.

File: Libplanet/Net/Protocols/RoutingTable.cs (hunk 15 above)
Comment (limebell) [reply Libplanet/Net/Protocols/RoutingTable.cs]: First time I thought this should be in a separated PR, but without this new test case fails. I'll add this change to the changelog.

File: Libplanet/Net/Messages/GetRecentStates.cs (hunk 12 above)
Comment (longfin) [reply Libplanet/Net/Messages/GetRecentStates.cs]: Oh, I see. how about reusing `this()` constructor?

**Comment by limebell** [comment]
Since broadcasting logic has changed, this patch is unnecessary.
